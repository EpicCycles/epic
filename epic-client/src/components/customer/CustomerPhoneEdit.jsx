import React from "react";
import {Icon} from "semantic-ui-react";
import FormTextInput from "../../common/FormTextInput";
import SelectInput from "../../common/SelectInput";
import {updateObject} from "../../helpers/utils";
import {addFieldToState, getModelKey, isModelValid} from "../app/model/helpers/model";
import {bikeFields, customerAddressFields, customerPhoneFields} from "../app/model/helpers/fields";
import EditModelRow from "../app/model/EditModelRow";

const newCustomerPhone = {
    number_type: 'H',
    telephone: '',
};
const initialState = {};

class CustomerPhoneEdit extends React.Component {
    state = initialState;

    componentWillMount() {
        this.setState({
            customerPhone: updateObject(this.props.customerPhone)
        });
    };

    handleInputChange = (fieldName, input) => {
        const updatedCustomerPhone = addFieldToState(this.state.customerPhone, customerPhoneFields, fieldName, input);
        this.setState({ customerPhone: updatedCustomerPhone });
    };

    onClickReset = () => {
        this.setState({
            customerPhone: updateObject(this.props.customerPhone)
        });
    };

    saveOrCreateCustomerPhone = () => {
        this.props.saveCustomerPhone(this.state.customerPhone);
    };

    onClickDelete = () => {
        if (this.props.customerPhone.id) {
            this.props.deleteCustomerPhone(this.props.customerPhone.id);
        } else {
            this.setState({customerPhone:newCustomerPhone});
        }
    };

    render() {
        const { customerPhone } = this.state;
        const { telephone, isChanged, } = customerPhone;
        const keyValue = getModelKey(customerPhone);
        const componentContext = keyValue;
        const rowClass = (customerPhone && customerPhone.error) ? "error" : "";
        const isValid = isModelValid(customerPhone);

        return <div className={rowClass} key={`row${componentContext}`}>
            <EditModelRow
                model={customerPhone}
                persistedModel={this.props.customerPhone}
                modelFields={customerPhoneFields}
                onChange={this.handleInputChange}
            />
            <div id={`actions_${componentContext}`}>
                  <span id={`actions${keyValue}`}>
                      {isChanged &&
                      <Icon id={`reset-phone${keyValue}`} name="undo"
                            onClick={this.onClickReset} title="Reset Phone details"
                      />
                      }
                      {isChanged &&
                      <Icon id={`accept-phone${keyValue}`} name="check" disabled={!isValid}
                            onClick={isValid && this.saveOrCreateCustomerPhone}
                            title="Confirm changes"/>
                      }
                      {telephone &&
                      <Icon id={`delete-phone${keyValue}`} name="delete"
                            onClick={this.onClickDelete}
                            title="Delete Phone Number"/>
                      }
                </span>
            </div>
        </div>;
    }
}

CustomerPhoneEdit.defaultProps = {
    customerPhone: newCustomerPhone,
};
export default CustomerPhoneEdit;

