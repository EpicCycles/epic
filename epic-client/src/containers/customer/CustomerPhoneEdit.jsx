import React from "react";
import {Icon} from "semantic-ui-react";
import FormTextInput from "../../common/FormTextInput";
import SelectInput from "../../common/SelectInput";

class CustomerPhoneEdit extends React.Component {
    state = {
        number_type: 'H',
        telephone: '',
        isChanged: false,
        isValid: true,
        options: [
            { value: 'H', text: 'Home', isDefault: true },
            { value: 'M', text: 'Mobile' },
            { value: 'W', text: 'Work' }
        ]
    };

    componentWillMount() {
        if (this.props.customerPhone) {
            this.setState({
                number_type: this.props.customerPhone.number_type,
                telephone: this.props.customerPhone.telephone
            });
        }
    };
    // TODO use componentDidUpdate(prevProps, prevState) to pick up change

    validateCustomerPhoneData = (number_type, telephone) => {
        let isChanged = false;
        let isValid = true;
        let telephoneError = "";

        if (this.props.customerPhone) {
            if (this.props.customerPhone.number_type !== number_type
                || this.props.customerPhone.telephone !== telephone) {
                isChanged = true;
            }
            if (!(telephone)) telephoneError = "Phone Number must be provided";
        } else {
            isChanged = !!telephone;
        }


        if (telephoneError) isValid = false;
        this.setState({
            number_type: number_type,
            telephone: telephone,
            isChanged: isChanged,
            telephoneError: telephoneError,
            isValid: isValid
        });
    };

    handleInputChange = (fieldName, input) => {
        if (fieldName.startsWith('telephone')) {
            this.validateCustomerPhoneData(this.state.number_type, input);
        }
        else {
            this.validateCustomerPhoneData(input, this.state.telephone);
        }
    };
    handleInputClear = (fieldName) => {
        if (fieldName.startsWith('removetelephone')) {
            const telephone = this.props.customerPhone ? this.props.customerPhone.telephone : '';
            this.validateCustomerPhoneData(this.state.number_type, telephone);
        }
    };

    onClickReset = () => {
        this.setState({
            telephone: this.props.customerPhone ? this.props.customerPhone.telephone : '',
            telephoneError: '',
            number_type: this.props.customerPhone ? this.props.customerPhone.number_type : 'H',
            isChanged: false,
            isValid: true
        });
    };

    saveOrCreateCustomerPhone = () => {
        if (this.props.customerPhone) {
            let phoneToSave = this.props.customerPhone;
            phoneToSave.telephone = this.state.telephone;
            phoneToSave.number_type = this.state.number_type;
            this.props.saveCustomerPhone(phoneToSave);
        }
        else {
            const newPhone = {
                customer: this.props.customerId,
                telephone: this.state.telephone,
                number_type: this.state.number_type
            };
            this.props.saveCustomerPhone(newPhone);
            this.setState({
                telephone: '',
                telephoneError: '',
                number_type: 'H',
                isChanged: false,
                isValid: true
            });
        }
    };

    onClickDelete = () => {
        if (this.props.customerPhone) {
            let phoneToSave = this.props.customerPhone;
            this.props.deleteCustomerPhone(phoneToSave.id);
        } else {
            this.setState({
                telephone: '',
                telephoneError: '',
                number_type: 'H',
                isChanged: false,
                isValid: true
            });
        }
    };

    render() {
        const { telephone, number_type, isChanged, isValid, telephoneError, options } = this.state;
        const { customerPhone } = this.props;
        const keyValue = (customerPhone && customerPhone.id) ? customerPhone.id : "new";
        const componentContext = customerPhone ? customerPhone.id : 'newPhone';
        const type_value = [number_type];
        return <tr id={componentContext}>
            <td>
                <SelectInput
                    fieldName={`number_type_${componentContext}`}
                    options={options}
                    onChange={this.handleInputChange}
                    value={type_value}
                />
            </td>
            <td>
                <FormTextInput
                    placeholder="Phone Number"
                    id={`telephone-input_${componentContext}`}
                    className="column full"
                    value={telephone}
                    fieldName={`telephone_${componentContext}`}
                    onChange={this.handleInputChange}
                    onClick={this.handleInputClear}
                    error={telephoneError}
                />
            </td>
            <td>
                {(customerPhone && customerPhone.add_date) ?
                    <span>Added on {customerPhone.add_date.substring(0, 10)}, last updated on {customerPhone.upd_date.substring(0, 10)}</span>
                    : <span>Add a new number</span>
                }
            </td>
            <td>
                  <span id={`actions${keyValue}`}>
                      {isChanged &&
                      <Icon id={`reset-phone${keyValue}`} name="undo"
                            onClick={this.onClickReset} title="Reset Phone details"
                      />
                      }
                      {isChanged &&
                      <Icon id={`accept-phone${keyValue}`} name="check" disabled={!isValid}
                            onClick={this.saveOrCreateCustomerPhone}
                            title="Confirm changes"/>
                      }
                      {telephone &&
                      <Icon id={`delete-phone${keyValue}`} name="delete"
                            onClick={this.onClickDelete}
                            title="Delete Phone Number"/>
                      }
                </span>
            </td>
        </tr>;
    }
}

export default CustomerPhoneEdit;

