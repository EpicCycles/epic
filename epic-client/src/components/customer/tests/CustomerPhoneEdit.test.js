import React from 'react';
import CustomerPhoneEdit from "../CustomerPhoneEdit";

// props - customerPhone {number_type, telephone}, customerId, saveCustomerPhone {function}, deleteCustomerPhone {function}
it("displays a passed customerPhone object correctly", () => {
    const customerPhone = {
        number_type: "H",
        telephone: "78736",
        id: 12
    };
    const component = shallow(
        <CustomerPhoneEdit customerPhone={customerPhone} />
    );
    expect(component).toMatchSnapshot();
});
it("displays a passed customerPhone object correctly when an error is present", () => {
    const customerPhone = {
        number_type: "H",
        telephone: "78736",
        id: 12,
        error:true
    };
    const component = shallow(
        <CustomerPhoneEdit customerPhone={customerPhone} />
    );
    expect(component).toMatchSnapshot();
});
it("sets the changed flag when a value changes", () => {
    const customerPhone = {
        number_type: "H",
        telephone: "78736",
        id: 12
    };
    const component = shallow(
        <CustomerPhoneEdit customerPhone={customerPhone}/>
    );
    component.instance().handleInputChange("telephone1", '12345678');
    expect(component.state('customerPhone').changed).toBeTruthy();

    component.instance().handleInputChange("telephone1", '78736');
    expect(component.state('customerPhone').changed).toBeTruthy();

    component.instance().handleInputChange("number_type1", 'M');
    expect(component.state('customerPhone').changed).toBeTruthy();
});
it("resets to passed state when reset if clicked", () => {
    const customerPhone = {
        number_type: "H",
        telephone: "78736",
        id: 12
    };
    const component = shallow(
        <CustomerPhoneEdit customerPhone={customerPhone}/>
    );
    expect(component.state('customerPhone')).toEqual(customerPhone);
    component.instance().handleInputChange("telephone1", "");
    expect(component.state('customerPhone').changed).toBeTruthy();
    component.instance().onClickReset();
    expect(component.state('customerPhone').changed).toBeFalsy();

});
it("calls method to save phone details and sets state as appropriate", () => {
    const saveCustomerPhone = jest.fn();
    const customerPhone = {
        number_type: "H",
        telephone: "78736",
        id: 12
    };
    const component = shallow(
        <CustomerPhoneEdit
            customerPhone={customerPhone}
            saveCustomerPhone={saveCustomerPhone}
        />
    );
    component.instance().handleInputChange("telephone1", '12345678');
    expect(component.state('customerPhone').changed).toBeTruthy();

    component.instance().handleInputChange("number_type1", 'M');
    expect(component.state('customerPhone').changed).toBeTruthy();

    component.instance().saveOrCreateCustomerPhone();
    expect(saveCustomerPhone.mock.calls.length).toBe(1);
    expect(saveCustomerPhone.mock.calls[0][0].number_type).toBe("M");
    expect(saveCustomerPhone.mock.calls[0][0].telephone).toBe("12345678");
    expect(saveCustomerPhone.mock.calls[0][0].id).toBe(customerPhone.id);
});
it("calls method to delete phone details and sets state as appropriate", () => {
    const deleteCustomerPhone = jest.fn();
    const customerPhone = {
        number_type: "H",
        telephone: "78736",
        id: 12
    };
    const component = shallow(
        <CustomerPhoneEdit
            customerPhone={customerPhone}
            deleteCustomerPhone={deleteCustomerPhone}
        />
    );

    component.instance().onClickDelete();
    expect(deleteCustomerPhone.mock.calls.length).toBe(1);
    expect(deleteCustomerPhone.mock.calls[0][0]).toBe(customerPhone.id);
});
it("calls method to save new phone details and sets state as appropriate", () => {
    const saveCustomerPhone = jest.fn();
    const customerPhone = {};
    const customerId = 23;

    const component = shallow(
        <CustomerPhoneEdit
            customerPhone={customerPhone}
            saveCustomerPhone={saveCustomerPhone}
            customerId={customerId}
        />
    );
    component.instance().handleInputChange("telephone1", '12345678');
    component.instance().handleInputChange("number_type1", 'M');

    component.instance().saveOrCreateCustomerPhone();
    expect(saveCustomerPhone.mock.calls.length).toBe(1);
    expect(saveCustomerPhone.mock.calls[0][0].number_type).toBe("M");
    expect(saveCustomerPhone.mock.calls[0][0].telephone).toBe("12345678");
});
it("calls handles clearing a new customer phone", () => {
    const deleteCustomerPhone = jest.fn();
    const customerPhone = {};
    const component = shallow(
        <CustomerPhoneEdit
            customerPhone={customerPhone}
            deleteCustomerPhone={deleteCustomerPhone}
        />
    );
    component.instance().handleInputChange("telephone1", '12345678');
    expect(component.state('customerPhone').changed).toBeTruthy();

    component.instance().handleInputChange("number_type1", 'M');
    expect(component.state('customerPhone').changed).toBeTruthy();

    component.instance().onClickDelete();
    expect(deleteCustomerPhone.mock.calls.length).toBe(0);
    expect(component.state('customerPhone').changed).toBeFalsy();
    expect(component.state('customerPhone').telephone).toBe("");
    expect(component.state('customerPhone').number_type).toBe('H');
});
