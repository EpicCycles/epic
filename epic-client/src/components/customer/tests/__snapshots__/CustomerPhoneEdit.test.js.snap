// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`displays a passed customerPhone object correctly 1`] = `
<tr
  className=""
  id={12}
>
  <td
    id="td1_12"
  >
    <SelectInput
      fieldName="number_type_12"
      onChange={[Function]}
      options={
        Array [
          Object {
            "isDefault": true,
            "text": "Home",
            "value": "H",
          },
          Object {
            "text": "Mobile",
            "value": "M",
          },
          Object {
            "text": "Work",
            "value": "W",
          },
        ]
      }
      value={
        Array [
          "H",
        ]
      }
    />
  </td>
  <td
    id="td2_12"
  >
    <FormTextInput
      className="column full"
      dataType="text"
      error=""
      fieldName="telephone_12"
      id="telephone-input_12"
      label=""
      onChange={[Function]}
      placeholder="Phone Number"
      size={30}
      title=""
      value="78736"
    />
  </td>
  <td
    id="td3_12"
  >
    <span
      id="spancomment_12"
    >
      Add a new number
    </span>
  </td>
  <td
    id="td4_12"
  >
    <span
      id="actions12"
    >
      <Icon
        as="i"
        id="delete-phone12"
        name="delete"
        onClick={[Function]}
        title="Delete Phone Number"
      />
    </span>
  </td>
</tr>
`;

exports[`displays a passed customerPhone object correctly when an error is present 1`] = `
<tr
  className="error"
  id={12}
>
  <td
    id="td1_12"
  >
    <SelectInput
      fieldName="number_type_12"
      onChange={[Function]}
      options={
        Array [
          Object {
            "isDefault": true,
            "text": "Home",
            "value": "H",
          },
          Object {
            "text": "Mobile",
            "value": "M",
          },
          Object {
            "text": "Work",
            "value": "W",
          },
        ]
      }
      value={
        Array [
          "H",
        ]
      }
    />
  </td>
  <td
    id="td2_12"
  >
    <FormTextInput
      className="column full"
      dataType="text"
      error=""
      fieldName="telephone_12"
      id="telephone-input_12"
      label=""
      onChange={[Function]}
      placeholder="Phone Number"
      size={30}
      title=""
      value="78736"
    />
  </td>
  <td
    id="td3_12"
  >
    <span
      id="spancomment_12"
    >
      Add a new number
    </span>
  </td>
  <td
    id="td4_12"
  >
    <span
      id="actions12"
    >
      <Icon
        as="i"
        id="delete-phone12"
        name="delete"
        onClick={[Function]}
        title="Delete Phone Number"
      />
    </span>
  </td>
</tr>
`;
